NGitLab.Tests/Docker/GitLabDockerContainer.cs

```diff
@@ -292,50 +316,112 @@ async Task GenerateAdminToken(GitLabCredential credentials)
+                             foreach (var element in form.Elements.OfType<IHtmlInputElement>().Where(e => e.Name == "personal_access_token[scopes][]"))
+                             {
+                                 element.IsChecked = true;
+                             }
+
+                             result = await form.SubmitAsync().ConfigureAwait(false);
+                             TestContext.Progress.WriteLine("Navigating to " + result.Location);
+                             if (result.StatusCode == HttpStatusCode.BadGateway)
+                                 throw new InvalidOperationException("Cannot generate a personal access token:\n" + result.ToHtml());
+
+                             var personalAccessTokenElement = result.GetElementById("created-personal-access-token");
+                             if (personalAccessTokenElement is null)
+                                 throw new InvalidOperationException("The page does not contain the element 'created-personal-access-token':\n" + result.ToHtml());
+
+                             credentials.AdminUserToken = personalAccessTokenElement.GetAttribute("value");
+                             return result;
+                         });
+                }
+
+                Task<IDocument> GenerateXProfileToken(GitLabCredential credentials, IBrowsingContext context, IDocument result)
+                {
+                    return Policy.Handle<InvalidOperationException>()
+                        .WaitAndRetryAsync(10, i => TimeSpan.FromSeconds(1))
+                        .ExecuteAsync(async () =>
+                        {
+                            TestContext.Progress.WriteLine("Generating request profiles token");
+                            result = await context.OpenAsync(GitLabUrl + "/admin/requests_profiles").ConfigureAwait(false);
+                            TestContext.Progress.WriteLine("Navigating to " + result.Location);
+                            if (result.StatusCode == HttpStatusCode.BadGateway)
+                                throw new InvalidOperationException("Cannot navigate to admin/requests_profiles:\n" + result.ToHtml());
+
+                            var codeElements = result.QuerySelectorAll("code").ToList();
+                            var tokenElement = codeElements.SingleOrDefault(n => n.TextContent.Trim().StartsWith("X-Profile-Token:", StringComparison.Ordinal));
+                            if (tokenElement == null)
+                                throw new InvalidOperationException("Cannot find X-Profile-Token in the page:\n" + result.ToHtml());
+
+                            credentials.ProfileToken = tokenElement.TextContent.Trim().Substring("X-Profile-Token:".Length).Trim();
+                            return result;
+                        });
+                }
+
+                Task<IDocument> SignIn(IBrowsingContext context)
```

Comment:
Can you define local functions in the order they are called, i.e.
- SignIn
- GeneratePersonalAccessToken
- GenerateXProfileToken

It'll make the code easier to follow.